api/v1beta1/frrconfiguration_webhook.go

```diff
@@ -0,0 +1,194 @@
+		Complete()
+}
+
+//+kubebuilder:webhook:verbs=create;update,path=/validate-frrk8s-metallb-io-v1beta1-frrconfiguration,mutating=false,failurePolicy=fail,groups=frrk8s.metallb.io,resources=frrconfigurations,versions=v1beta1,name=frrconfigurationsvalidationwebhook.metallb.io,sideEffects=None,admissionReviewVersions=v1
+
+var _ webhook.Validator = &FRRConfiguration{}
+
+type nodeAndConfigs struct {
+	name   string
+	labels map[string]string
+	cfgs   *FRRConfigurationList
+}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for FRRConfiguration.
+func (frrConfig *FRRConfiguration) ValidateCreate() error {
+	level.Debug(Logger).Log("webhook", "frrconfiguration", "action", "create", "name", frrConfig.Name, "namespace", frrConfig.Namespace)
+
+	if frrConfig.Namespace != Namespace {
+		return fmt.Errorf("resource must be created in %s namespace", Namespace)
+	}
+
+	selector, err := metav1.LabelSelectorAsSelector(&frrConfig.Spec.NodeSelector)
+	if err != nil {
+		return errors.Wrapf(err, "resource contains an invalid NodeSelector")
+	}
+
+	existingNodes, err := getExistingNodes()
+	if err != nil {
+		return err
+	}
+
+	existingFRRConfigurations, err := getExistingFRRConfigurations()
+	if err != nil {
+		return err
+	}
+
+	matchingNodes := []nodeAndConfigs{}
+	for _, n := range existingNodes.Items {
+		if selector.Matches(labels.Set(n.Labels)) {
+			matchingNodes = append(matchingNodes, nodeAndConfigs{
+				name:   n.Name,
+				labels: n.Labels,
+				cfgs:   &FRRConfigurationList{},
+			})
+		}
+	}
+
+	for _, n := range matchingNodes {
+		for _, cfg := range existingFRRConfigurations.Items {
+			nodeSelector := cfg.Spec.NodeSelector
+			selector, err := metav1.LabelSelectorAsSelector(&nodeSelector)
+			if err != nil {
+				// shouldn't happen as it would have been denied earlier, just in case.
+				continue
+			}
+
+			if cfg.Name == frrConfig.Name {
+				// shouldn't happen as it would be considered an update, just in case as we add
+				// the "newer" one at the end anyways.
+				continue
+			}
+
+			if selector.Matches(labels.Set(n.labels)) {
+				n.cfgs.Items = append(n.cfgs.Items, *cfg.DeepCopy())
+			}
+		}
+		n.cfgs.Items = append(n.cfgs.Items, *frrConfig.DeepCopy())
+	}
+
+	for _, n := range matchingNodes {
+		err := Validate(n.cfgs)
+		if err != nil {
+			level.Error(Logger).Log("webhook", "frrconfiguration", "action", "create", "name", frrConfig.Name, "namespace", frrConfig.Namespace, "node", n.name, "error", err)
+			return errors.Wrapf(err, "resource is invalid for node %s", n.name)
+		}
+	}
+
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for FRRConfiguration.
+func (frrConfig *FRRConfiguration) ValidateUpdate(old runtime.Object) error {
+	level.Debug(Logger).Log("webhook", "frrconfiguration", "action", "update", "name", frrConfig.Name, "namespace", frrConfig.Namespace)
+
+	selector, err := metav1.LabelSelectorAsSelector(&frrConfig.Spec.NodeSelector)
```

Comment:
This is more or less the same as ValidateCreate, can we factor them together?

Reply:
done